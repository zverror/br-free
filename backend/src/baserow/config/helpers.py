import asyncio

from loguru import logger


class dummy_context:
    async def __aenter__(self):
        pass

    async def __aexit__(self, exc_type, exc, traceback):
        pass


class ConcurrencyLimiterASGI:
    """
    Helper wrapper on ASGI app to limit the number of requests handled
    at the same time.
    """

    def __init__(self, app, max_concurrency: int | None = None):
        self.app = app
        logger.info(f"Setting ASGI app concurrency to {max_concurrency}")
        self.semaphore = (
            asyncio.Semaphore(max_concurrency)
            if (isinstance(max_concurrency, int) and max_concurrency > 0)
            else dummy_context()
        )

    async def __call__(self, scope, receive, send):
        async with self.semaphore:
            await self.app(scope, receive, send)
