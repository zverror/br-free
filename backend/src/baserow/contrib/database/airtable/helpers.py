from typing import List

from baserow.contrib.database.fields.models import Field, SelectOption

from .constants import AIRTABLE_BASEROW_COLOR_MAPPING


def import_airtable_date_type_options(type_options) -> dict:
    date_format_mapping = {"European": "EU", "US": "US"}
    time_format_mapping = {"12hour": "12", "24hour": "24"}
    return {
        "date_format": date_format_mapping.get(type_options.get("dateFormat"), "ISO"),
        "date_include_time": type_options.get("isDateTime", False),
        "date_time_format": time_format_mapping.get(
            type_options.get("timeFormat"), "24"
        ),
    }


def import_airtable_choices(field_id: str, type_options: dict) -> List[SelectOption]:
    order = type_options.get("choiceOrder", [])
    choices = type_options.get("choices", [])
    return [
        SelectOption(
            # Combine select id with choice id as choice id is not guaranteed to be
            # unique across table
            id=f"{field_id}_{choice['id']}",
            value=choice["name"],
            color=AIRTABLE_BASEROW_COLOR_MAPPING.get(
                # The color isn't always provided, hence the fallback to an empty
                # color, which will fallback on the blue color.
                choice.get("color", ""),
                "blue",
            ),
            order=order.index(choice["id"]),
        )
        for choice in choices.values()
    ]


def set_select_options_on_field(
    field: Field, field_id: str, type_options: dict
) -> Field:
    """
    Set the `select_options` of a field in the prefetched objects cache. Overriding
    this object makes sure that when later `field.select_options.all()` is executed,
    it will return Airtable choices. This for example happens in
    `FieldType::export_serialized`.

    :param field: The field where the select must be set on.
    :param field_id: id of airtable choice field
    :param type_options: The options where to extract the Airtable choices from.
    :return: The updated field object.
    """

    # Overriding this object makes sure that when later
    # `field.select_options.all()` is executed, it will return Airtable choices.
    # This for example happens in `FieldType::export_serialized`.
    field._prefetched_objects_cache = {
        "select_options": import_airtable_choices(field_id, type_options)
    }
    return field
