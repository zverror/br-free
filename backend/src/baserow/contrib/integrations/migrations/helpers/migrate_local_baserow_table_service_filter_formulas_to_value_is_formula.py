"""
Helpers for 0006_migrate_local_baserow_table_service_filter_formulas_to_value_is_formula
"""

from baserow.core.formula import BaserowFormulaSyntaxError, get_parse_tree_for_formula


def value_parses_as_formula(value: str) -> bool:
    """
    Returns whether the service filter value is a formula or not.
    """

    # If the value is empty, it's not a formula.
    if not value:
        return False

    # If it's a string that's an integer, it's not a formula.
    try:
        int(value)
        return False
    except ValueError:
        pass

    try:
        get_parse_tree_for_formula(value)
        return True
    except BaserowFormulaSyntaxError:
        return False
