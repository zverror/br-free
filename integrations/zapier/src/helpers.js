const { unsupportedBaserowFieldTypes } = require('./constants')

/**
 * Fetches the fields of a table and converts them to an array with valid Zapier
 * field objects.
 */
const getRowInputValues = async (z, bundle) => {
  if (!bundle.inputData.tableID) {
    throw new Error('The `tableID` must be provided.')
  }

  const fieldsGetRequest = await z.request({
    url: `${bundle.authData.apiURL}/api/database/fields/table/${bundle.inputData.tableID}/`,
    method: 'GET',
    headers: {
      'Accept': 'application/json',
      'Authorization': `Token ${bundle.authData.apiToken}`,
    },
  })

  return fieldsGetRequest.json.map(v => {
    return mapBaserowFieldTypesToZapierTypes(v)
  })
}

/**
 * Fetches the fields and converts the input data to Baserow row compatible data.
 */
const prepareInputDataForBaserow = async (z, bundle) => {
  if (!bundle.inputData.tableID) {
    throw new Error('The `tableID` must be provided.')
  }

  const fieldsGetRequest = await z.request({
    url: `${bundle.authData.apiURL}/api/database/fields/table/${bundle.inputData.tableID}/`,
    method: 'GET',
    headers: {
      'Accept': 'application/json',
      'Authorization': `Token ${bundle.authData.apiToken}`,
    },
  })

  let rowData = { id: bundle.inputData.rowID }
  fieldsGetRequest
    .json
    .filter(
      (baserowField) =>
        baserowField.read_only
          || !unsupportedBaserowFieldTypes.includes(baserowField.type)
    )
    .filter((baserowField) => bundle.inputData.hasOwnProperty(baserowField.name))
    .forEach(baserowField => {
      let value = bundle.inputData[baserowField.name]

      if (baserowField.type === 'multiple_collaborators') {
        value = value.map(id => {
          return { id }}
        )
      }

      rowData[baserowField.name] = value
    })

  return rowData
}

/**
 * Converts the provided Baserow field type object to a Zapier compatible object.
 */
const mapBaserowFieldTypesToZapierTypes = (baserowField) => {
  const zapType = {
    key: baserowField.name,
    label: baserowField.name,
    type: 'string'
  }

  if (baserowField.type === 'long_text') {
    zapType.type = 'text'
  }

  if (baserowField.type === 'boolean') {
    zapType.type = 'boolean'
  }

  if (baserowField.type === 'number') {
    zapType.type = 'integer'

    if (baserowField.number_decimal_places > 0) {
      zapType.type = 'float'
    }
  }

  if (baserowField.type === 'boolean') {
    zapType.type = 'boolean'
  }

  if (baserowField.type === 'rating') {
    zapType.type = 'integer'
  }

  if (['single_select', 'multiple_select'].includes(baserowField.type)) {
    const choices = {}
    baserowField.select_options.forEach(el => {
      choices[`${el.id}`] = el.value
    })
    zapType.type = 'string'
    zapType.choices = choices
  }

  if (baserowField.type === 'multiple_select') {
    zapType.list = true
  }

  if (baserowField.type === 'link_row') {
    zapType.type = 'integer'
    zapType.helpText = 'Provide row ids that you want to link to.'
    zapType.list = true
  }

  if (baserowField.type === 'multiple_collaborators') {
    zapType.type = 'integer'
    zapType.helpText = 'Provide user ids that you want to link to.'
    zapType.list = true
  }

  if (baserowField.type === 'date' && !baserowField.date_include_time) {
    zapType.type = 'date'
    zapType.helpText =
      'the date fields accepts a date in ISO format (e.g. 2020-01-01)'
  }

  if (baserowField.type === 'date' && baserowField.date_include_time) {
    zapType.type = 'datetime'
    zapType.helpText =
      'the date fields accepts date and time in ISO format (e.g. 2020-01-01 12:00)'
  }

  if (
    baserowField.read_only
    || unsupportedBaserowFieldTypes.includes(baserowField.type)
  ) {
    // Read only and the file field are not supported.
    return
  }

  return zapType
}

module.exports = {
  getRowInputValues,
  prepareInputDataForBaserow,
  mapBaserowFieldTypesToZapierTypes,
}
