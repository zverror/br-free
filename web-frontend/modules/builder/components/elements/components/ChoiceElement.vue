<template>
  <ABFormGroup
    :label="labelResolved"
    :required="element.required"
    :error-message="displayFormDataError ? $t('error.requiredField') : ''"
    :style="getStyleOverride('input')"
  >
    <ABDropdown
      v-if="element.show_as_dropdown"
      v-model="inputValue"
      class="choice-element"
      :placeholder="
        canHaveOptions ? placeholderResolved : $t('choiceElement.addOptions')
      "
      :show-search="false"
      :multiple="element.multiple"
      @hide="onFormElementTouch"
    >
      <ABDropdownItem
        v-for="option in optionsResolved"
        :key="option.id"
        :name="option.name || (option.value ? option.value : '')"
        :value="option.value"
      />
    </ABDropdown>
    <template v-else>
      <template v-if="canHaveOptions">
        <template v-if="element.multiple">
          <ABCheckbox
            v-for="option in optionsResolved"
            :key="option.id"
            :read-only="isEditMode"
            :error="displayFormDataError"
            :value="inputValue.includes(option.value)"
            @input="onOptionChange(option, $event)"
          >
            {{ option.name || option.value }}
          </ABCheckbox>
        </template>
        <template v-else>
          <ABRadio
            v-for="option in optionsResolved"
            :key="option.id"
            :read-only="isEditMode"
            :error="displayFormDataError"
            :value="option.value === inputValue"
            @input="onOptionChange(option, $event)"
          >
            {{ option.name || option.value }}
          </ABRadio>
        </template>
      </template>
      <template v-else>{{ $t('choiceElement.addOptions') }}</template>
    </template>
  </ABFormGroup>
</template>

<script>
import formElement from '@baserow/modules/builder/mixins/formElement'
import {
  ensureString,
  ensureStringOrInteger,
  ensureArray,
} from '@baserow/modules/core/utils/validator'
import { CHOICE_OPTION_TYPES } from '@baserow/modules/builder/enums'

export default {
  name: 'ChoiceElement',
  mixins: [formElement],
  props: {
    /**
     * @type {Object}
     * @property {string} label - The label displayed above the choice element
     * @property {string} default_value - The default value selected
     * @property {string} placeholder - The placeholder value of the choice element
     * @property {boolean} required - If the element is required for form submission
     * @property {boolean} multiple - If the choice element allows multiple selections
     * @property {boolean} show_as_dropdown - If the choice element should be displayed as a dropdown
     * @property {Array} options - The options of the choice element
     * @property {string} option_type - The type of the options
     * @property {string} formula_name - The formula for the name of the option
     * @property {string} formula_value - The formula for the value of the option
     */
    element: {
      type: Object,
      required: true,
    },
  },
  computed: {
    labelResolved() {
      return ensureString(this.resolveFormula(this.element.label))
    },
    placeholderResolved() {
      return ensureString(this.resolveFormula(this.element.placeholder))
    },
    defaultValueResolved() {
      if (this.element.multiple) {
        const existingValues = this.optionsResolved.map(({ value }) => value)
        return ensureArray(this.resolveFormula(this.element.default_value))
          .map(ensureStringOrInteger)
          .filter((value) => existingValues.includes(value))
      } else {
        // Always return a string if we have a default value, otherwise
        // set the value to null as single select fields will only skip
        // field preparation if the value is null.
        const resolvedSingleValue = ensureStringOrInteger(
          this.resolveFormula(this.element.default_value)
        )

        return resolvedSingleValue === '' ? null : resolvedSingleValue
      }
    },
    canHaveOptions() {
      return !this.elementIsInError
    },
    optionsResolved() {
      switch (this.element.option_type) {
        case CHOICE_OPTION_TYPES.MANUAL:
          return this.element.options.map(({ name, value }) => ({
            name,
            value: value === null ? name : value,
          }))
        case CHOICE_OPTION_TYPES.FORMULAS: {
          const formulaValues = ensureArray(
            this.resolveFormula(this.element.formula_value)
          )
          const formulaNames = ensureArray(
            this.resolveFormula(this.element.formula_name)
          )
          return formulaValues.map((value, index) => ({
            id: index,
            value: ensureStringOrInteger(value),
            name: ensureString(
              index < formulaValues.length ? formulaNames[index] : value
            ),
          }))
        }
        default:
          return []
      }
    },
  },
  watch: {
    defaultValueResolved: {
      handler(newValue) {
        this.inputValue = newValue
      },
      immediate: true,
    },
    'element.multiple'() {
      this.setFormData(this.defaultValueResolved)
    },
  },
  methods: {
    onOptionChange(option, value) {
      if (value) {
        if (this.element.multiple) {
          this.inputValue = [...this.inputValue, option.value]
        } else {
          this.inputValue = option.value
        }
      } else if (this.element.multiple) {
        this.inputValue = this.inputValue.filter((v) => v !== option.value)
      }
    },
  },
}
</script>
