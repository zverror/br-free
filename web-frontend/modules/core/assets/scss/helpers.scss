%clearfix {
  &::after {
    clear: both;
    content: '';
    display: table;
  }
}

.clearfix {
  @extend %clearfix;
}

.hidden {
  display: none;
}

.visibility-hidden {
  visibility: hidden;
  transform: translateX(-10000px) translateY(-10000px);
}

.align-right {
  text-align: right;
}

.background-white {
  background-color: $white;
}

.color-primary {
  color: $color-primary-500 !important;
}

.color-success {
  color: $color-success-500 !important;
}

.color-warning {
  color: $color-warning-500 !important;
}

.color-error {
  color: $color-error-500 !important;
}

.color-neutral {
  color: $color-neutral-500 !important;
}

.padding-top-2 {
  padding-top: 16px !important;
}

.margin-top-0 {
  margin-top: 0 !important;
}

.margin-top-1 {
  margin-top: 8px !important;
}

.margin-top-2 {
  margin-top: 16px !important;
}

.margin-top-3 {
  margin-top: 30px !important;
}

.margin-top-4 {
  margin-top: 40px !important;
}

.margin-bottom-0 {
  margin-bottom: 0 !important;
}

.margin-bottom-1 {
  margin-bottom: 8px !important;
}

.margin-bottom-2 {
  margin-bottom: 16px !important;
}

.margin-bottom-24 {
  margin-top: 24px !important;
}

.margin-bottom-3 {
  margin-bottom: 30px !important;
}

.margin-bottom-4 {
  margin-bottom: 40px !important;
}

.margin-left-auto {
  margin-left: auto !important;
}

.margin-left-0 {
  margin-left: 0 !important;
}

.margin-left-1 {
  margin-left: 8px !important;
}

.margin-left-2 {
  margin-left: 16px !important;
}

.margin-left-3 {
  margin-left: 30px !important;
}

.margin-left-4 {
  margin-left: 40px !important;
}

.margin-right-auto {
  margin-right: auto !important;
}

.margin-right-0 {
  margin-right: 0 !important;
}

.margin-right-1 {
  margin-right: 8px !important;
}

.margin-right-2 {
  margin-right: 16px !important;
}

.margin-right-3 {
  margin-right: 30px !important;
}

.margin-right-4 {
  margin-right: 40px !important;
}

.resizing-horizontal {
  cursor: col-resize;
}

// This selector can only be used to temporarily show an element for calculating
// positions.
.forced-block {
  display: block !important;
}

// Sometimes we forcefully want to disable the user-select property to initial
// because the element is for example being edited via a contenteditable attribute
// which does not work in combination with the user-select: none; property in Safari.
.forced-user-select-initial {
  user-select: initial !important;
}

.forced-text-overflow-initial {
  text-overflow: initial !important;
}

.forced-pointer-events-auto {
  pointer-events: auto !important;
}

.prevent-scroll {
  overflow: hidden !important;
}

.whitespace-pre-wrap {
  white-space: pre-wrap;
}

.remove-number-input-controls,
%remove-number-input-controls {
  appearance: textfield !important;

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    appearance: none;
    margin: 0;
  }
}

.flex {
  width: 100%;
  display: flex;
  gap: var(--gap, 10px);
}

.flex-wrap {
  flex-wrap: wrap;
}

.flex-100 {
  width: 100%;
  flex-basis: 100%;
}

.flex-grow-1 {
  flex-grow: 1;
}

.align-items-center {
  align-items: center;
}

.justify-content-space-between {
  justify-content: space-between;
}

.position-relative {
  position: relative;
}

.inline-block {
  display: inline-block;
}

@keyframes spin {
  0% {
    transform: rotate(0);
  }

  50% {
    transform: rotate(360deg);
  }

  100% {
    transform: rotate(720deg);
  }
}

/* spacer classes generation */

@each $size in $spacings {
  .p-#{$size},
  %p-#{$size} {
    padding: #{$size}px;
  }

  .px-#{$size},
  %px-#{$size} {
    padding-left: #{$size}px;
    padding-right: #{$size}px;
  }

  .py-#{$size},
  %py-#{$size} {
    padding-top: #{$size}px;
    padding-bottom: #{$size}px;
  }

  .m-#{$size},
  %m-#{$size} {
    margin: #{$size}px;
  }

  .my-#{$size},
  %my-#{$size} {
    margin-top: #{$size}px;
    margin-bottom: #{$size}px;
  }

  .mx-#{$size},
  %mx-#{$size} {
    margin-left: #{$size}px;
    margin-right: #{$size}px;
  }

  .pl-#{$size},
  %pl-#{$size} {
    padding-left: #{$size}px;
  }

  .ml-#{$size},
  %ml-#{$size} {
    margin-left: #{$size}px;
  }

  .pr-#{$size},
  %pr-#{$size} {
    padding-right: #{$size}px;
  }

  .mr-#{$size},
  %mr-#{$size} {
    margin-right: #{$size}px;
  }

  .pb-#{$size},
  %pb-#{$size} {
    padding-bottom: #{$size}px;
  }

  .mb-#{$size},
  %mb-#{$size} {
    margin-bottom: #{$size}px;
  }

  .pt-#{$size},
  %pt-#{$size} {
    padding-top: #{$size}px;
  }

  .mt-#{$size},
  %mt-#{$size} {
    margin-top: #{$size}px;
  }
}
