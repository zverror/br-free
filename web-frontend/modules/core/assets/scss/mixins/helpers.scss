@mixin fixed-height($height, $font-size) {
  height: $height;
  line-height: $height;
  font-size: $font-size;
}

@mixin center-text($width, $font-size, $height: $width) {
  @include fixed-height($height, $font-size);

  width: $width;
  text-align: center;
}

@mixin flex-align-items($gap: 0, $display: flex) {
  display: $display;
  align-items: center;
  gap: $gap;
}

@mixin loading($size: 1.4rem) {
  position: relative;
  display: block;
  width: $size;
  height: $size;
  border-radius: 50%;
  border: 0.25em solid;
  border-color: $color-primary-500 transparent $color-primary-500 transparent;
  animation: spin infinite 1800ms;
  animation-timing-function: cubic-bezier(0.785, 0.135, 0.15, 0.86);
}

// The basically works the same as padding: 10px 0 0 10px or padding: 10px; only then
// resulting position: absolute; with corresponding top, right, bottom, left values.
// -------
// Example:
//  @include absolute(10px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 10px; bottom: 10px; left: 10px;
// -------
// Example:
//  @include absolute(10px, 20px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 10px; left: 20px;
// -------
// Example:
//  @include absolute(10px, 20px, 30px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 30px;
@mixin absolute($top: null, $right: null, $bottom: null, $left: null) {
  position: absolute;

  @if $top != null and $right == null and $bottom == null and $left == null {
    inset: $top $top $top $top;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else if $top != null and $right != null and $bottom == null and $left == null
  {
    inset: $top $right $top $right;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else {
    @if $top {
      top: $top;
    }

    @if $right {
      right: $right;
    }

    @if $bottom {
      bottom: $bottom;
    }

    @if $left {
      left: $left;
    }
  }
}
