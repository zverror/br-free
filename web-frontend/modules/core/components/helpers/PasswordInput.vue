<!--
This helper component can only be used in a form which uses the password validation
from modules/core/validators.js
as ist expects the following vuelidate validation state params

- required
- maxLength
- minLength

In order to use this component in another component/form it needs a value
prop which can be passed via the 'v-model' directive. That means the state for the 
password value always lives in the parent component.
The parent component also needs to provide the vuelidate password validation in order
for this component to show all the possible validation errors.

Example Usage:

<PasswordInput
   v-model="PASSWORDSTATE"
   :validation-state="the vuelidate validation instance of the given password value"
/>

In case the password validation rules change, the passwordValidation object in the
validators file needs to be updated, as well as this component.

modules/core/validators.js

-->
<template>
  <FormGroup
    small-label
    :label="label"
    required
    :error="validationState.$error"
  >
    <FormInput
      :error="validationState.$error"
      type="password"
      size="large"
      :autocomplete="autocomplete"
      :value="value"
      :placeholder="placeholder"
      @blur="validationState.$touch()"
      @input="$emit('input', $event)"
    >
    </FormInput>

    <template #error>
      <span v-if="validationState.$error && !validationState.required">
        <i v-if="showErrorIcon" class="iconoir-warning-triangle"></i>
        {{ $t('error.inputRequired') }}
      </span>
      <span v-if="validationState.$error && !validationState.maxLength">
        <i v-if="showErrorIcon" class="iconoir-warning-triangle"></i>
        {{
          $t('error.maxLength', {
            max: validationState.$params.maxLength.max,
          })
        }}
      </span>
      <span v-if="validationState.$error && !validationState.minLength">
        <i v-if="showErrorIcon" class="iconoir-warning-triangle"></i>
        {{
          $t('error.minLength', {
            min: validationState.$params.minLength.min,
          })
        }}
      </span>
    </template>
  </FormGroup>
</template>

<script>
export default {
  name: 'PasswordInput',
  props: {
    validationState: {
      type: Object,
      required: true,
    },
    value: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      required: false,
      default: null,
    },
    autocomplete: {
      type: String,
      required: false,
      default: 'new-password',
    },
    placeholder: {
      type: String,
      required: false,
      default: null,
    },
    errorPlaceholderClass: {
      type: String,
      required: false,
      default: '',
    },
    showErrorIcon: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
}
</script>
