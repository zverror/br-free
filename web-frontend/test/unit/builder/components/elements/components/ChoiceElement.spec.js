import { TestApp } from '@baserow/test/helpers/testApp'
import ChoiceElement from '@baserow/modules/builder/components/elements/components/ChoiceElement.vue'

describe('ChoiceElement', () => {
  let testApp = null
  let store = null

  beforeAll(() => {
    testApp = new TestApp()
    store = testApp.store
  })

  afterEach(() => {
    testApp.afterEach()
  })

  const mountComponent = ({ props = {}, slots = {}, provide = {} }) => {
    return testApp.mount(ChoiceElement, {
      propsData: props,
      slots,
      provide,
    })
  }

  const mountComponentForElement = async (element) => {
    const builder = { id: 1, theme: { primary_color: '#ccc' } }
    const page = { id: 1, elements: [] }
    const workspace = {}
    const mode = 'public'
    const applicationContext = { builder, page, mode }

    store.dispatch('element/forceCreate', { page, element })

    const elementType = store.$registry.get('element', 'choice')

    const defaultValue = element.multiple ? [] : null
    const payload = {
      value: defaultValue,
      type: elementType.formDataType(element),
      isValid: elementType.isValid(element, defaultValue, applicationContext),
      touched: false,
    }

    store.dispatch('formData/setFormData', {
      page,
      elementId: element.id,
      payload,
    })

    const wrapper = await mountComponent({
      props: {
        element,
      },
      provide: {
        builder,
        currentPage: page,
        elementPage: page,
        mode,
        applicationContext: { builder, page, mode },
        element,
        workspace,
      },
    })
    return wrapper
  }

  test('as default', async () => {
    const wrapper = await mountComponentForElement({
      id: 42,
      defaultValue: '1',
      parent_element_id: null,
      type: 'choice',
      multiple: false,
      option_type: 'manual',
      show_as_dropdown: true,
      options: [],
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('as manual dropdown', async () => {
    const wrapper = await mountComponentForElement({
      id: 42,
      defaultValue: '1',
      parent_element_id: null,
      type: 'choice',
      multiple: false,
      option_type: 'manual',
      show_as_dropdown: true,
      options: [
        { value: '1', name: 'First' },
        { value: '2', name: 'Second' },
      ],
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('as manual dropdown - values', async () => {
    const wrapper = await mountComponentForElement({
      id: 42,
      defaultValue: '1',
      parent_element_id: null,
      type: 'choice',
      multiple: false,
      option_type: 'manual',
      show_as_dropdown: true,
      options: [
        { value: '', name: 'Foo Name' },
        { value: 'bar_name', name: 'Bar Name' },
        { value: null, name: 'Baz Name' },
      ],
    })

    expect(wrapper.vm.optionsResolved).toEqual([
      // An empty string is a valid Value
      { value: '', name: 'Foo Name' },
      // 'bar_name' is a valid Value
      { value: 'bar_name', name: 'Bar Name' },
      // null is replaced by the name, i.e. 'Baz Name'
      { value: 'Baz Name', name: 'Baz Name' },
    ])
  })

  test('as manual radio', async () => {
    const wrapper = await mountComponentForElement({
      id: 42,
      defaultValue: '1',
      parent_element_id: null,
      type: 'choice',
      multiple: false,
      option_type: 'manual',
      show_as_dropdown: false,
      options: [
        { value: '1', name: 'First' },
        { value: '2', name: 'Second' },
      ],
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('as manual checkboxes', async () => {
    const wrapper = await mountComponentForElement({
      id: 42,
      defaultValue: '1',
      parent_element_id: null,
      type: 'choice',
      multiple: true,
      option_type: 'manual',
      show_as_dropdown: false,
      options: [
        { value: '1', name: 'First' },
        { value: '2', name: 'Second' },
      ],
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
