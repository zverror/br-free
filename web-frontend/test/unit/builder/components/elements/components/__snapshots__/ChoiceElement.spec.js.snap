// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChoiceElement as default 1`] = `
<div
  class="control ab-form-group"
>
  <!---->
   
  <!---->
   
  <div
    class="control__wrapper"
  >
    <div
      class="control__elements"
    >
      <div
        class="flex-grow-1"
      >
        <!---->
         
        <div
          class="ab-form-group__children"
        >
          <div
            class="ab-dropdown choice-element"
            tabindex="0"
          >
            <a
              class="ab-dropdown__selected"
            >
              <span
                class="ab-dropdown__selected-placeholder"
              >
                choiceElement.addOptions
              </span>
               
              <i
                class="ab-dropdown__toggle-icon iconoir-nav-arrow-down"
              />
            </a>
             
            <div
              class="ab-dropdown__items hidden"
            >
              <!---->
               
              <ul
                class="select__items"
                style="display: none;"
                tabindex="-1"
              />
               
              <div
                class="select__items--empty"
              >
                
        dropdown.empty
      
              </div>
               
              <!---->
            </div>
          </div>
           
          <!---->
        </div>
      </div>
       
      <!---->
       
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`ChoiceElement as manual checkboxes 1`] = `
<div
  class="control ab-form-group"
>
  <!---->
   
  <!---->
   
  <div
    class="control__wrapper"
  >
    <div
      class="control__elements"
    >
      <div
        class="flex-grow-1"
      >
        <!---->
         
        <div
          class="ab-form-group__children"
        >
          <div
            class="ab-checkbox"
          >
            <input
              class="ab-checkbox__input"
              type="checkbox"
            />
             
            <label
              class="ab-checkbox__label"
            >
              
          First
        
            </label>
          </div>
          <div
            class="ab-checkbox"
          >
            <input
              class="ab-checkbox__input"
              type="checkbox"
            />
             
            <label
              class="ab-checkbox__label"
            >
              
          Second
        
            </label>
          </div>
           
          <!---->
        </div>
      </div>
       
      <!---->
       
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`ChoiceElement as manual dropdown 1`] = `
<div
  class="control ab-form-group"
>
  <!---->
   
  <!---->
   
  <div
    class="control__wrapper"
  >
    <div
      class="control__elements"
    >
      <div
        class="flex-grow-1"
      >
        <!---->
         
        <div
          class="ab-form-group__children"
        >
          <div
            class="ab-dropdown choice-element"
            tabindex="0"
          >
            <a
              class="ab-dropdown__selected"
            >
              <span
                class="ab-dropdown__selected-placeholder"
              >
                action.makeChoice
              </span>
               
              <i
                class="ab-dropdown__toggle-icon iconoir-nav-arrow-down"
              />
            </a>
             
            <div
              class="ab-dropdown__items hidden"
            >
              <!---->
               
              <ul
                class="select__items"
                style=""
                tabindex="-1"
              >
                <li
                  class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                >
                  <a
                    class="ab-dropdownitem__item-link"
                  >
                    <div
                      class="ab-dropdownitem__item-name"
                    >
                      <!---->
                       
                      <span
                        class="ab-dropdownitem__item-name-text"
                        title="First"
                      >
                        First
                      </span>
                    </div>
                     
                    <!---->
                  </a>
                   
                  <i
                    class="ab-dropdownitem__item-active-icon iconoir-check"
                  />
                </li>
                <li
                  class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                >
                  <a
                    class="ab-dropdownitem__item-link"
                  >
                    <div
                      class="ab-dropdownitem__item-name"
                    >
                      <!---->
                       
                      <span
                        class="ab-dropdownitem__item-name-text"
                        title="Second"
                      >
                        Second
                      </span>
                    </div>
                     
                    <!---->
                  </a>
                   
                  <i
                    class="ab-dropdownitem__item-active-icon iconoir-check"
                  />
                </li>
              </ul>
               
              <!---->
               
              <!---->
            </div>
          </div>
           
          <!---->
        </div>
      </div>
       
      <!---->
       
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`ChoiceElement as manual radio 1`] = `
<div
  class="control ab-form-group"
>
  <!---->
   
  <!---->
   
  <div
    class="control__wrapper"
  >
    <div
      class="control__elements"
    >
      <div
        class="flex-grow-1"
      >
        <!---->
         
        <div
          class="ab-form-group__children"
        >
          <div
            class="ab-radio"
          >
            <input
              class="ab-radio__input"
              type="radio"
            />
             
            <label
              class="ab-radio__label"
            >
              
          First
        
            </label>
          </div>
          <div
            class="ab-radio"
          >
            <input
              class="ab-radio__input"
              type="radio"
            />
             
            <label
              class="ab-radio__label"
            >
              
          Second
        
            </label>
          </div>
           
          <!---->
        </div>
      </div>
       
      <!---->
       
    </div>
     
    <!---->
  </div>
</div>
`;
